crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -12,48 +12,315 @@
+                    OperationType::Health | OperationType::Custom(_) => Ok((vec![], None)),
+
+                    OperationType::Restart => {
+                        context
+                            .publish_restart_operation_status(entity, &cmd_id, message)
+                            .await
+                    }
+                    OperationType::SoftwareList => {
+                        context
+                            .publish_software_list(entity, &cmd_id, &message)
+                            .await
+                    }
+                    OperationType::SoftwareUpdate => {
+                        context
+                            .publish_software_update_status(entity, &cmd_id, &message)
+                            .await
+                    }
+                    OperationType::LogUpload => {
+                        context
+                            .handle_log_upload_state_change(entity, &cmd_id, &message)
+                            .await
+                    }
+                    OperationType::ConfigSnapshot => {
+                        context
+                            .handle_config_snapshot_state_change(entity, &cmd_id, &message)
+                            .await
+                    }
+                    OperationType::ConfigUpdate => {
+                        context
+                            .handle_config_update_state_change(entity, &cmd_id, &message)
+                            .await
+                    }
+                    OperationType::FirmwareUpdate => {
+                        context
+                            .handle_firmware_update_state_change(entity, &cmd_id, &message)
+                            .await
+                    }
+                };
+
+                let mut mqtt_publisher = context.mqtt_publisher.clone();
+                match res {
+                    // If there are mapped final status messages to be published, they are cached until the operation
+                    // log is uploaded
+                    Ok((messages, command)) => {
+                        if let Some(command) = command {
+                            if let Err(e) = context
+                                .upload_operation_log(&external_id, &cmd_id, &operation, command)
+                                .await
+                            {
+                                error!("failed to upload operation logs: {e}");
+                            }
+                        }
+
+                        for message in messages {
+                            // if task publishes MQTT clearing message, an operation is considered finished, so we can
+                            // terminate the task as well
+                            if message.retain
+                                && message.payload_bytes().is_empty()
+                                && message.topic == command_topic
+                            {
+                                rx.close();
```

Comment:
Minor: Just like the `upload_operation_log` is done above, at one place, for all operations, generating and publishing the clear messages also could be done from here, without expecting all the different operation handlers to do the same (resulting in duplicated code). But this can be deferred for a follow-up PR.

Reply:
will be done in a follow-up: https://github.com/thin-edge/thin-edge.io/issues/2880#issuecomment-2223213567